### PR TITLE
Additional gradle tests

### PR DESCRIPTION
**Pull request by kmfas**
Added some new tests for the gradle plugin.
Tests now use a fixed Version of gradle instead of using the last available one.
Refactored some code and added comments.

### CODE REVIEW

File: src/test/resources/gradle_plugin/pipeline_test_build_scan.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+stage ('build_Project') {
+    steps {
+        script {
+           if(isUnix()) {
```
Comment (uhafner) [inline src/test/resources/gradle_plugin/pipeline_test_build_scan.txt]: I needed to replace lines 4-10 with the following to get it running. The path hack below does not work on my machine.

```
        withEnv(["PATH+GRADLE=${tool 'Gradle 4.0'}/bin"]) {
            sh "gradle hello --scan"
        }
```

File: src/main/java/org/jenkinsci/test/acceptance/plugins/gradle/GradleTask.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package org.jenkinsci.test.acceptance.plugins.gradle;
+
+/**
+ * Created by wlaschinger on 29.06.17.
```
Comment (uhafner) [inline src/main/java/org/jenkinsci/test/acceptance/plugins/gradle/GradleTask.java]: Please update your file template

@author Vorname Nachname

File: src/main/java/org/jenkinsci/test/acceptance/plugins/gradle/GradleTask.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ */
+public enum  GradleTask {
+
+    HELLO("hello", "Hello world!"),
+    FIRST("firstTask", "First!"),
+    SECOND("secondTask", "Second!"),
+    ENVIRONMENT_VARIABLES("environmentVariables", null),
```
Comment (uhafner) [inline src/main/java/org/jenkinsci/test/acceptance/plugins/gradle/GradleTask.java]: Using null as a parameter is a code smell: better introduce additional one-arg constructor.

File: src/test/java/plugins/GradlePluginTest.java [hunk 4]
```diff
@@ -28,103 +28,160 @@
...
 import static org.jenkinsci.test.acceptance.Matchers.containsString;
+import static org.jenkinsci.test.acceptance.plugins.gradle.GradleTask.HELLO;
+import static org.jenkinsci.test.acceptance.plugins.gradle.GradleTask.FIRST;
+import static org.jenkinsci.test.acceptance.plugins.gradle.GradleTask.SECOND;
+import static org.jenkinsci.test.acceptance.plugins.gradle.GradleTask.ENVIRONMENT_VARIABLES;
+import static org.jenkinsci.test.acceptance.plugins.gradle.GradleTask.JOB_PARAM_AS_PROJECT_PROPERTIES;
+import static org.jenkinsci.test.acceptance.plugins.gradle.GradleTask.JOB_PARAM_AS_SYSTEM_PROPERTIES;
 import static org.junit.Assert.assertThat;
 
 @WithPlugins("gradle")
```
Comment (uhafner) [inline src/test/java/plugins/GradlePluginTest.java]: If you like feel free to add a JavaDoc with @author

File: src/test/java/plugins/GradlePluginTest.java [hunk 5]
```diff
@@ -135,93 +192,274 @@ public void run_gradle_script_build_scan_link() {
...
+
+    /**
+     * Jenkins File to run multiple gradle tasks
+     */
+    private static final String JENKINS_FILE_MULTIPLE_TASKS = "/gradle_plugin/pipeline_test_multiple_tasks.txt";
+
+    /**
+     * Runs a pipeline gradle build and verifies the build was executed successfully
+     */
+    @Test
```
Comment (uhafner) [inline src/test/java/plugins/GradlePluginTest.java]: All pipeline tests require `@WithPlugins("workflow-aggregator")`

File: src/test/resources/gradle_plugin/pipeline_test_build_scan.txt (hunk 1 above)
Comment (wolfs) [inline src/test/resources/gradle_plugin/pipeline_test_build_scan.txt]: As of Gradle plugin `2.17.1` setting the path shouldn't be necessary if the tool is already selected. I think this PR should have fixed it: https://github.com/jenkinsci/gradle-plugin/pull/42

Comment (uhafner) [inline src/test/resources/gradle_plugin/pipeline_test_build_scan.txt]: But the tool command is required, or not? Otherwise grade will not be installed.

Comment (wolfs) [inline src/test/resources/gradle_plugin/pipeline_test_build_scan.txt]: That is my understanding: When using the tool command then it is automatically added to the path.

Comment (kmfas) [inline src/test/resources/gradle_plugin/pipeline_test_build_scan.txt]: Sorry i didn't have time to take a look at this earlier.
Using gradle now as described in: https://jenkins.io/doc/book/pipeline/syntax/#tools
Thanks for the additional input!

File: src/test/java/plugins/GradlePluginTest.java (hunk 4 above)
Comment (kmfas) [inline src/test/java/plugins/GradlePluginTest.java]: Didn't add any.

File: src/test/java/plugins/GradlePluginTest.java [hunk 6]
```diff
@@ -135,93 +192,274 @@ public void run_gradle_script_build_scan_link() {
...
+        job.save();
+
+        final Build build = job.startBuild().shouldSucceed();
+        assertThat(build.getConsole(), containsString("Gradle Properties: hello world"));
+    }
+
+    /**
+     * Verify the existence of links for executed tasks.
+     */
+    @Test @WithPlugins("gradle@1.27-SNAPSHOT")
```
Comment (olivergondza) [inline src/test/java/plugins/GradlePluginTest.java]: Why `1.27-SNAPSHOT`?

File: src/main/tool_installers/updates/hudson.plugins.gradle.GradleInstaller [hunk 7]
```diff
@@ -1,307 +1 @@
...
    "id": "0.8",
    "name": "Gradle 0.8",
    "url": "https://services.gradle.org/distributions/gradle-0.8-bin.zip"
  },
    {
    "id": "0.7",
    "name": "Gradle 0.7",
    "url": "https://services.gradle.org/distributions/gradle-0.7-bin.zip"
  }
]}
```
Comment (olivergondza) [inline src/main/tool_installers/updates/hudson.plugins.gradle.GradleInstaller]: Please preserve the formatting if you can to minimize the diff size.

File: src/test/java/plugins/GradlePluginTest.java (hunk 6 above)
Comment (uhafner) [inline src/test/java/plugins/GradlePluginTest.java]: As far as I remember the versioning of the plugin is broken, see https://issues.jenkins-ci.org/browse/JENKINS-45126. Maybe a question for @wolfs

Comment (wolfs) [inline src/test/java/plugins/GradlePluginTest.java]: @uhafner 1.27.1 works just fine. 1.27 has been removed from the update center.

Comment (kmfas) [inline src/test/java/plugins/GradlePluginTest.java]: The Versioning was somewhat broken but should work now.
The Annotations should not be needed anymore.
I thought, they did not do any harm so i left them in. I can remove them if you want.

File: src/main/tool_installers/updates/hudson.plugins.gradle.GradleInstaller (hunk 7 above)
Comment (kmfas) [inline src/main/tool_installers/updates/hudson.plugins.gradle.GradleInstaller]: If you mean the order of the entries, this is the same behaviour the plugin/tool installation itself has. At some point/version the order changed.
Does it make sense to reorder the entries manually?

**Pull request merged**